src/shared/env.ts

```diff
@@ -2,30 +2,39 @@ import { isBrowser } from './utils';
 
 const testHost = 'localhost:8536';
 
-const internalHost =
+let internalHost =
   (!isBrowser() && process.env.LEMMY_INTERNAL_HOST) || testHost; // used for local dev
-export const externalHost = isBrowser()
-  ? `${window.location.hostname}${
-      ['1234', '1235'].includes(window.location.port)
-        ? ':8536'
-        : window.location.port == ''
-        ? ''
-        : `:${window.location.port}`
-    }`
-  : process.env.LEMMY_EXTERNAL_HOST || testHost;
-
-const secure = isBrowser()
-  ? window.location.protocol == 'https:'
-    ? 's'
-    : ''
-  : process.env.LEMMY_HTTPS == 'true'
-  ? 's'
-  : '';
-
-const host = isBrowser() ? externalHost : internalHost;
+export let externalHost: string;
+let host: string;
+let wsHost: string;
+let secure: string;
+
+if (isBrowser()) {
+  // browser
+  const lemmyConfig =
+    typeof window.lemmyConfig !== 'undefined' ? window.lemmyConfig : {};
+
+  externalHost = `${window.location.hostname}${
+    ['1234', '1235'].includes(window.location.port)
+      ? ':8536'
+      : window.location.port == ''
+      ? ''
+      : `:${window.location.port}`
+  }`;
+
+  host = externalHost;
+  wsHost = lemmyConfig.wsHost || host;
+  secure = window.location.protocol == 'https:' ? 's' : '';
+} else {
+  // server-side
+  externalHost = process.env.LEMMY_EXTERNAL_HOST || testHost;
+  host = internalHost;
+  wsHost = process.env.LEMMY_WS_HOST || host;
+  secure = process.env.LEMMY_HTTPS == 'true' ? 's' : '';
+}
```

Comment:
My main problem with this is that the actual host, both the http and websocket, are not the issue here. They both should not be pointing at anything but the real host. 

Your issue is that you want some of the static resources, and bundles, to use your own custom CDN. That should be done in the `server/index.tsx`. You've tried to get around this issue by making the CDN your "real host", and then adding a config to workaround that.

Reply:
CloudFront (the CDN) is the "real host" because that's what the DNS that someone types into the browser should resolve to.
CloudFront does a fair bit of what one would do with nginx, except you don't have to run anything. It handles caching, reverse proxying, SSL termination, etc. It's a good practice to put it in front of everything, although it is no good for WS. 
I get your point that it should be used for just the static files, but it needs to be what the browser talks to. 

IMO the fewer containers and services I have to run and the more I can offload onto AWS the better. That's generally how this cloud stuff works; it's not just a datacenter it's a set of managed services you can configure and get some great benefits vs. managing your own services. 

I didn't get to federating yet so I'm unclear on why nginx is needed to federate.

Reply:
Nginx is 100% needed for federating. And cloudfront is not running the lemmy server.

Reply:
I'm okay with having my infrastructure on AWS. Maybe you are thinking of CloudFlare?

Reply:
Ah sry you're right lol. This is new territory because I have no idea how you'd write the nginx rerouting rules we have for AWS. But I think this should mostly be done through that. I

Lemmy's backend includes both an http and websocket API, and I don't think extricating them is the problem you're trying to solve. You're trying to figure out how to get static resources, served up by your CDN. Those are included in `server/index.tsx`.

Reply:
What I'm trying to do in this PR is simply make the websocket host optionally configurable. People may have good reasons to want to do this, such as separating out the services hosting the frontend and backend, or load balancing WS (network load balancers) differently from HTTP (application load balancers) for example. 

In my case I would rather do my rewrite rules with cloud-native infrastructure, because it's one less server for me to run, has caching at 225 POPs, handles SSL termination, supports HTTP/2, compression, IPv6, DNS alias integration, and a lot of other wonderful features.
I'm looking at https://raw.githubusercontent.com/LemmyNet/lemmy/main/ansible/templates/nginx.conf right now, is this the config that is needed to run lemmy and federation properly? I think these rules are quite translatable to Cloudfront. What I've been working on is trying to translate the configuration needed to run lemmy into a reusable IaC definition for AWS so anyone can deploy a highly available lemmy instance with the minimum amount of effort.

Reply:
Nginx is definitely not required for federating, but you will have to translate the nginx config for your provider. Mainly the parts which go to the backend (everything with `{{ lemmy_port }}` in that template).

Reply:
Ok I've done that here https://github.com/jetbridge/lemmy-cdk/blob/master/lib/lemmy/loadbalancer.ts#L103

Reply:
I'll test this locally, but I have reservations about merging this. Its essentially supporting an amazon-specific limitation since they don't support websockets, and you have to offload that work to its own container, whilst amazon does everything else? Websockets is the main way lemmy-ui communicates, so it seems odd to have AWS handle everything BUT the main way lemmy communicates.

Reply:
There's nothing AWS-specific here, just making a hostname configurable. Other people may have other setups where they want different hostnames or ports for frontend/backend. The small refactoring and ability to pass through configuration from environment -> server.js -> browser will probably come in handy for future frontend configuration options as well.